openadapt/share.py

```diff
@@ -0,0 +1,120 @@
+
+
+def export_recording_to_folder(recording_id: int) -> None:
+    """Export a recording to a zip file.
+
+    Args:
+        recording_id (int): The ID of the recording to export.
+
+    Returns:
+        str: The path of the created zip file.
+    """
+    # TODO: replace with alternative, e.g.:
+    # - https://stackoverflow.com/questions/28238785/sqlalchemy-python-multiple-databases
+    # - https://github.com/bigbag/sqlalchemy-multiple-db
+    recording_db_path = db.export_recording(recording_id)
+
+    if recording_db_path:
+        # Create the directory if it doesn't exist
+        os.makedirs(config.ZIPPED_RECORDING_FOLDER_PATH, exist_ok=True)
+
+        # Path to the source db file
+        db_filename = f"recording_{recording_id}.db"
+
+        # Path to the compressed file
+        zip_filename = f"recording_{recording_id}.zip"
+        zip_path = os.path.join(config.ZIPPED_RECORDING_FOLDER_PATH, zip_filename)
+
+        # Create an in-memory zip file and add the db file
+        with ZipFile(zip_path, "w", ZIP_DEFLATED, compresslevel=9) as zip_file:
+            zip_file.write(recording_db_path, arcname=db_filename)
+
+        logger.info(f"created {zip_path=}")
+
+        # delete db file
+        os.remove(recording_db_path)
+        logger.info(f"deleted {recording_db_path=}")
+
+        return zip_path
+
+
+def send_file(file_path: str) -> None:
+    """Send a file using the 'wormhole' command-line tool.
+
+    Args:
+        file_path (str): The path of the file to send.
+    """
+    # Construct the command
+    command = ["wormhole", "send", file_path]
+
+    # Execute the command
+    try:
+        subprocess.run(command, check=True)
+    except subprocess.CalledProcessError as e:
+        raise Exception(f"Error occurred while running 'wormhole send': {e}")
+
+
+def send_recording(recording_id: int) -> None:
+    """Export a recording to a zip file and send it to another computer.
+
+    Args:
+        recording_id (int): The ID of the recording to send.
+    """
+    zip_file_path = export_recording_to_folder(recording_id)
+
+    if zip_file_path:
+        try:
+            send_file(zip_file_path)
+            # File sent successfully
+        except Exception as exc:
+            logger.exception(exc)
+        finally:
+            # Delete the zip file after sending or in case of exception
+            if os.path.exists(zip_file_path):
+                os.remove(zip_file_path)
+                logger.info(f"deleted {zip_file_path=}")
+
+
+def receive_recording(wormhole_code: str) -> None:
+    """Receive a recording zip file from another computer using a wormhole code.
+
+    Args:
+        wormhole_code (str): The wormhole code to receive the recording.
+    """
+    # Construct the command
+    command = ["wormhole", "receive", wormhole_code]
+
+    # Execute the command
+    try:
+        subprocess.run(command, check=True)
+    except subprocess.CalledProcessError as exc:
+        logger.exception(exc)
```

Comment:
To close the loop, I think we want to:

1. extract the .db file into the data/recordings directory (so that it's in the same relative location as it was before it was zipped)
2. optionally visualize it, e.g. via command line argument --visualize

This might also require modifying visualize.py to take a different database file

Reply:
Currently when exporting I only copy the recording data into the recording table for a given db. But if I want to visualize it I'll need to copy all the contents relating to the recording. So I would copy all the ActionEvents and other tables and filter by timestamp(As highlighted in screenshot --> i printed all data in db file).
 
![image](https://github.com/OpenAdaptAI/OpenAdapt/assets/43456930/6842f337-9e0d-4aa7-9e9a-f5a011fcd548)

Reply:
@abrichr Addressed in https://github.com/OpenAdaptAI/OpenAdapt/pull/223/commits/2f1548d31aab0fe4010cf46241b4695f04428bd7. Now it copies all data for a recording filtered by the recording_timestamp. This will include migrations, so ideally when visualizing a new db, we would alembic upgrade head for it work.

Reply:
I've broken down the steps into both following:
1. for extracting db file
- [x] pass in recording_id in receieve recording function
- [x] extract db file in data/recordings

2. visualize
- [x] create new fire function with --visualize with recording_id parameter
- [x] create helper function to create new engine and bind to new db in data/recordings
- [x] run the visualizer